### PR TITLE
Fix bug 1063152 - add change control

### PR DESCRIPTION
**Pull request by jwhitlock**
`django.contrib.auth` is used to add action and resource API permissions, such as the ability to add a Browser, change an existing Browser, or delete a Browser.  Groups are used to gather common permissions, such as the "change-resource" group to allow add and changes of API resources.  The `user` representation is expanded to include the permissions groups as well as link to `changesets` the user has contributed.  The browse app now exposes the `section`, `user`, and `changeset` resources (but not historical resources yet).

[browsercompat.herokuapp.com](http://browsercompat.herokuapp.com/) has been updated with this code.

This commit builds on PR #22  and PR #23.  Once those are merged, this code can be rebased so it only includes the relevant commits for [bug 1063152](https://bugzilla.mozilla.org/show_bug.cgi?id=1063152).

### CODE REVIEW

**Comment by jwhitlock** [comment]
Rebased

**Comment by jwhitlock** [comment]
Rebased after PR #23 merged

**Comment by jwhitlock** [comment]
I'm researching why the Travis build failed.  I suspect it is a pip 6.0.8 vs 1.5.6 on the django-simple-history requirement:

    git+git://github.com/jwhitlock/django-simple-history.git@order_wrt_included#egg=django-simple-history

If there are changes, I may refresh the requirements again.

**Comment by jwhitlock** [comment]
The issue was with my fork of drf-json-api:

    https://github.com/jwhitlock/drf-json-api/archive/v0.1d.zip

I switched it to the editable requirements version:

    git+git://github.com/jwhitlock/drf-json-api@v0.1d#egg=drf-json-api

I took the opportunity to update other requirements, including a Django security update.  Django REST Framework is red, but it is uncertain if the fix will be back ported to 2.4.x (https://github.com/tomchristie/django-rest-framework/pull/2755)

**Comment by jwhitlock** [comment]
Travis CI is doing something different this week than last week.  Even when a dependency is pinned in requirements.txt, it is installing the unpinned, latest file.  I'll investigate why Travis is behaving differently than locally running `tox`.

**Comment by groovecoder** [comment]
Weird, we might be hitting that with MDN too? ... https://github.com/travis-ci/travis-ci/issues/3436

**Comment by jwhitlock** [comment]
Ug, looks like an issue with tox 1.9.1, found and fixed today: https://bitbucket.org/hpk42/tox/issue/230/tox-191-ignores-package-versions-specified

New push pins to tox 1.9.0 until 1.9.2 is released.

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
+    - TOXENV=docs
+    - TOXENV=flake8
+    - TOXENV=coverage DATABASE_URL="postgres://postgres@localhost:5432/travis_ci_test"
 
-python:
-  - "3.4"
-  - "2.7"
-
-install: pip install -r requirements.txt coveralls
+install: pip install tox==1.9.0
```
Comment (groovecoder) [inline .travis.yml]: 1.9.2 is now released and should also fix the error. :neutral_face:

File: .travis.yml [hunk 2]
```diff
@@ -1,12 +1,13 @@
 # Config file for automatic testing at travis-ci.org
 
 language: python
+env:
+    - TOXENV=py27 DATABASE_URL="postgres://postgres@localhost:5432/travis_ci_test"
+    - TOXENV=py34 DATABASE_URL="postgres://postgres@localhost:5432/travis_ci_test"
+    - TOXENV=docs
+    - TOXENV=flake8
+    - TOXENV=coverage DATABASE_URL="postgres://postgres@localhost:5432/travis_ci_test"
```
Comment (jezdez) [inline .travis.yml]: You can set global env vars with the "global" keyword: http://docs.travis-ci.com/user/environment-variables/#Global-Variables

File: requirements.txt [hunk 3]
```diff
@@ -16,25 +16,20 @@ django-filter==0.9.2
...
 # JSON API interfaces for Django REST Framework
-https://github.com/jwhitlock/drf-json-api/archive/v0.1d.zip
+git+git://github.com/jwhitlock/drf-json-api@v0.1d#egg=drf-json-api
 
 # History of changes to models
-# django-simple-history==1.5.0 has deprecation warnings w/ Django 1.7
-# PR #141 - Fix issue with Meta.order_with_respect_to
-# Use my django-simple-history w/ PR #141 on 2014-11-16
-git+git://github.com/jwhitlock/django-simple-history.git@order_wrt_included#egg=django-simple-history
+django-simple-history==1.5.4
```
Comment (groovecoder) [inline requirements.txt]: :smiley: for getting back to main repo.

File: webplatformcompat/tests/base.py [hunk 4]
```diff
@@ -22,13 +22,27 @@ def reverse(self, viewname, **kwargs):
         """Create a full URL for a view"""
         return self.baseUrl + reverse(viewname, kwargs=kwargs)
 
-    def login_superuser(self):
-        """Create and login a superuser, saving to self.user"""
+    def login_user(self, groups=None):
+        """Create and login a user, saving to self.user.
+
+        groups - default groups for new users.  Defaults to ['change-resource']
```
Comment (groovecoder) [inline webplatformcompat/tests/base.py]: This doesn't seem to be true? I.e., the default param is `groups=None` and the rest of the tests explicitly pass the `['change-resource']` as needed.

**Comment by groovecoder** [comment]
The test code is growing duplicitous.  E.g., in this PR there are 7 distinct `test_post_not_allowed` tests that do exactly the same logic against 7 different urls. At the very least, this logic could be moved up into a method in `base.APITestCase`. 

Do you find the viewset tests to be brittle? They seem to test the entire integration of custom wpc code (e.g., `HistoricalModelSerializer`), DRF serialization wiring (e.g., `BrowserSerializer.Meta.fields`), and default DRF viewset logic. Have they been a pain to maintain as we make changes?

File: webplatformcompat/tests/base.py (hunk 4 above)
Comment (jwhitlock) [inline webplatformcompat/tests/base.py]: The function default `None` is turned into `['change-resource]` on line 39.  I'm avoiding the issues with using a list as the default parameter: http://effbot.org/zone/default-values.htm .  So, that's the disconnect between the comment and the function signature

Also, I was looking at the tests - none of them pass `['change-resource']`.  Are you saying you'd be more comfortable if all the callers were explicit?  I was trying to avoid changing many lines when the default user permissions changes.

I'm open to a change, just not sure what you are suggesting.

File: .travis.yml (hunk 1 above)
Comment (jwhitlock) [inline .travis.yml]: I will unpin.  We may lose comments on later commits.

**Comment by jwhitlock** [comment]
I'm mixing integration tests with unit tests.  I think one "user in default group POSTs to a resource requiring additional permissions" is a good unit test, and one is sufficient for all similar resources, "user in default group POSTs to resource type X" is a good integration test, and you'd need an integration test for each resource.  Would you be more comfortable if I split out the integration tests into a separate folder structure?  Is that part of this PR?

I'll agree on cleaning up the `test_post_not_allowed` tests for this round.  It's premature testing - I expect the writing team to request finer grained permissions once the API ships.  Those permissions would be executed with database migrations, and the changes would only appear as changing tests, not changing code.

The end-to-end API tests fail on requirements updates - Django 1.6 to 1.7, Django REST framework 2.3 to 2.4, or drf-json-api 0.1.0 to 0.1.1.  DRF is a big bucket of hacks, and will probably continue to break code on updates until at least 3.3.

File: .travis.yml (hunk 2 above)
Comment (jwhitlock) [inline .travis.yml]: Is this what you are thinking @jezdez? https://github.com/jwhitlock/web-platform-compat/blob/f4d6f4c7f0b7a0309243e0a69a874810c011fdef/.travis.yml

Comment (jezdez) [inline .travis.yml]: @jwhitlock Yup!

**Comment by jwhitlock** [comment]
Changes in this latest update:
- tox 1.9.2 plays nice with requirements.txt
- Refactor travis.yml to global + matrix environment, per @jezdez 
- Add migration for django-simple-history 1.5.4 model updates
- Remove redundant integration tests, leaving the PUT, POST, and DELETE tests on the browser view, per @groovecoder 

https://browsercompat.herokuapp.com is up-to-date with these changes (plus migration).
